lib/src/clojure_lsp/handlers.clj

```diff
@@ -67,31 +68,48 @@
                :delay/start delay-start
                :delay/end now})))))))
 
-(defn wait-for-changes [uri db*]
-  (wait-for-all-changes [uri] db*))
-
 (defmacro logging-delayed-task [delay-data task-id & body]
-  (let [process-msg (str task-id " %s")
-        wait-and-process-msg (str process-msg " - waited %s")
-        timeout-msg (format "Timeout in %s waiting for changes to %%s" task-id)
+  (let [cancelled-msg (str task-id " cancelled - waited %s")
+        timed-out-msg (format "Timeout in %s waiting for changes to %%s" task-id)
+        immediate-msg (str task-id " %s")
+        waited-msg (str immediate-msg " - waited %s")
         msg-sym (gensym "log-message")]
     `(let [delay-data# ~delay-data
            delay-outcome# (:delay/outcome delay-data#)]
-       (if (= :timed-out delay-outcome#)
-         (let [~msg-sym (format ~timeout-msg (first (:delay/timeout-uris delay-data#)))]
+       (case delay-outcome#
+         :cancelled
+         (let [~msg-sym (format ~cancelled-msg (shared/format-time-delta-ms (:delay/start delay-data#) (:delay/end delay-data#)))]
+           ~(with-meta `(logger/debug ~msg-sym) (meta &form)))
+         :timed-out
+         (let [~msg-sym (format ~timed-out-msg (first (:delay/timeout-uris delay-data#)))]
            ~(with-meta `(logger/warn ~msg-sym) (meta &form)))
          (let [result# (do ~@body)
                ~msg-sym (case delay-outcome#
                           :immediate
-                          (format ~process-msg
+                          (format ~immediate-msg
                                   (shared/start-time->end-time-ms (:delay/start delay-data#)))
                           :waited
-                          (format ~wait-and-process-msg
+                          (format ~waited-msg
                                   (shared/start-time->end-time-ms (:delay/end delay-data#))
                                   (shared/format-time-delta-ms (:delay/start delay-data#) (:delay/end delay-data#))))]
            ~(with-meta `(logger/info ~msg-sym) (meta &form))
            result#)))))
 
+(defmacro process-after-all-changes
+  [components uris task & body]
+  (with-meta
+    `(logging-delayed-task
+       (wait-for-changes ~components ~uris)
+       ~task
+       ~@body)
+    (meta &form)))
```

Comment:
I was about to suggest moving both calls to a single function in the previous PR, glad you noticed that as well

Reply:
👍 That PR also removed the name `process-after-changes`, and this restores it. I'm glad to have it back—it's a useful shorthand for this part of the code.